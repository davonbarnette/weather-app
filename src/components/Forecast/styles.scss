@import "../../styles/global";

.forecast {
  @include flex-common(column, flex-start, center);

  .chart {
    @include flex-common(row, flex-start, flex-start);
    width:100%;
    margin-bottom:12px;
  }

  .days {
    @include flex-common(row, space-between, center);
    padding:0 12px;
    width:100%;

    .single-date {
      @include flex-common(column, center, center);
      color:$dark-gray;
      font-weight:600;
      border: solid 1px #e0e0e0;
      border-radius:3px;
      padding:12px;
      transition:all 0.2s;

      .temps {
        @include flex-common(row, space-between, center);
        font-size:13px;
        font-weight:500;
        margin-top: 12px;
        .min-temp {
          color:$light-gray;
          margin-right:9px;
        }
        .max-temp {
          color:$dark-gray;
        }
      }
      &:not(.selected):hover {
        cursor: pointer;
      }

      &.selected {
        background:#f3f3f3;
      }
    }
  }

}