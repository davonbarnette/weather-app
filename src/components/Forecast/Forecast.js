import React, {useEffect, useState} from 'react';
import {AreaChart} from 'react-chartkick';
import ReactMoment from 'react-moment';
import moment from 'moment';
import cx from 'classnames';
import 'chart.js';

import './styles.scss';

const Forecast = (props) => {
    const { list } = props.forecast;

    // We can assume the selectedDay is the first day on the list.
    let startingDay = getISODate(getLocalDateFromDT(list[0].dt), true);
    const [selectedDay, setSelectedDay] = useState(startingDay);
    const [forecastByDay, setForecastByDay] = useState(undefined);

    useEffect(()=> {

        // Upon updating the forecast list, we make a key:value map of the days keyed to the list. This way, we won't
        // have to iterate over the forecast list every time someone clicks a different day.
        let newForecastByDay = {};
        props.forecast.list.forEach((threeHourForecast) => {
            let date = getISODate(getLocalDateFromDT(threeHourForecast.dt), true);
            if (!newForecastByDay[date]) newForecastByDay[date] = {
                list: [],
            };
            newForecastByDay[date].list.push(threeHourForecast);
            if (!newForecastByDay[date].minTemp || threeHourForecast.main.temp < newForecastByDay[date].minTemp){
                newForecastByDay[date].minTemp = threeHourForecast.main.temp;
            }
            if (!newForecastByDay[date].maxTemp || threeHourForecast.main.temp > newForecastByDay[date].maxTemp){
                newForecastByDay[date].maxTemp = threeHourForecast.main.temp;
            }
        })
        setForecastByDay(newForecastByDay);
    }, [props.forecast])

    function getAreaChartData(){
        if (!forecastByDay) return undefined;
        let data = {};
        let list = forecastByDay[selectedDay].list;
        list.forEach(listItem => {
            data[getLocalDateFromDT(listItem.dt).format('h:mm A')] = listItem.main.temp;
        });
        return data;
    }

    function getISODate(date, withoutTime){
        let iso = date.toISOString();
        if (withoutTime) return iso.split('T')[0];
        else return iso;
    }

    function getLocalDateFromDT(dt){
        return moment((dt) * 1000);
    }

    function onDateClick(selectedDay){
        setSelectedDay(selectedDay);
    }

    let libraryConfiguration = {
        scales: {
            gridLines: {
                display:false,
            }
        }
    }

    return (
        <section className='forecast'>
            <div className='chart'>
                <AreaChart library={libraryConfiguration} data={getAreaChartData()} width='100%'/>
            </div>
            <div className='days'>
                {forecastByDay && Object.keys(forecastByDay).map(key => {
                    let day = forecastByDay[key];
                    let firstInterval = day.list[0]
                    if (firstInterval){
                        let date = getLocalDateFromDT(firstInterval.dt)
                        let selected = selectedDay === key;
                        return (
                            <div key={key} className={cx('single-date', {selected})} onClick={()=>onDateClick(key)}>
                                <ReactMoment local date={date} format={'ddd'}/>
                                <div className='temps'>
                                    <div className='min-temp'>{day.minTemp.toFixed(0)}&#8457;</div>
                                    <div className='max-temp'>{day.maxTemp.toFixed(0)}&#8457;</div>
                                </div>
                            </div>
                        )
                    }
                    else return null;
                })}
            </div>
        </section>
    )
}

export default Forecast;