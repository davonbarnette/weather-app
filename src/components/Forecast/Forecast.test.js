import React from 'react';
import Forecast from "./Forecast";
import { render } from '@testing-library/react';

describe('Forecast Component', () => {
    test('Snapshot', () => {
        let props = {
            forecast:{
                city: {timezone: -14400},
                list: [
                    {dt: 1593442800, dt_text: "2020-06-29 15:00:00", main: {temp: 80.51}}
                ]
            }
        }
        const {container} = render(<Forecast {...props}/>)
        expect(container.firstChild).toMatchSnapshot();

    })
})