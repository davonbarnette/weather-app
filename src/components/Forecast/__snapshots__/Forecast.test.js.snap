// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast Component Snapshot 1`] = `
<section
  class="forecast"
>
  <div
    class="chart"
  >
    <div
      id="chart-1"
      style="height: 300px; line-height: 300px; width: 100%; text-align: center; color: rgb(153, 153, 153); font-size: 14px; font-family: 'Lucida Grande', 'Lucida Sans Unicode', Verdana, Arial, Helvetica, sans-serif;"
    >
      <canvas />
    </div>
  </div>
  <div
    class="days"
  >
    <div
      class="single-date selected"
    >
      <time
        datetime="1593442800000"
      >
        Mon
      </time>
      <div
        class="temps"
      >
        <div
          class="min-temp"
        >
          81
          ℉
        </div>
        <div
          class="max-temp"
        >
          81
          ℉
        </div>
      </div>
    </div>
  </div>
</section>
`;
